### PR TITLE
csclient: add Login method and make Upload* use it

### PR DESCRIPTION
**Pull request by rogpeppe**
This means that uploads can work even when they're large.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/901/
Test PASSed.

File: csclient/csclient.go [hunk 1]
```diff
@@ -531,3 +548,25 @@ func (cs *Client) Log(typ params.LogType, level params.LogLevel, message string,
...
+// for the charm store and stores them in the client's
+// cookie jar.
+func (cs *Client) Login() error {
+	var m macaroon.Macaroon
+	if err := cs.Get("/macaroon", &m); err != nil {
+		return errgo.Notef(err, "cannot retrieve the authentication macaroon")
+	}
+	ms, err := httpbakery.DischargeAll(&m, cs.params.HTTPClient, cs.params.VisitWebPage)
+	if err != nil {
+		return errgo.Mask(err)
```
Comment (urosj) [inline csclient/csclient.go]: I wonder if need more descriptive errors ... but i guess that can be done on higher level within the charmcmd.

File: csclient/csclient_test.go [hunk 2]
```diff
@@ -1196,3 +1189,40 @@ func (s *suite) TestMacaroonAuthorization(c *gc.C) {
 	c.Assert(result.IdRevision.Revision, gc.Equals, rurl.URL.Revision)
 	c.Assert(httpbakery.IsInteractionError(errgo.Cause(err)), gc.Equals, true)
 }
+
+func (s *suite) TestLogin(c *gc.C) {
+	rurl := newResolvedURL("~charmers/utopic/wordpress-42", 42)
+	err := s.store.AddCharmWithArchive(rurl, storetesting.Charms.CharmDir("wordpress"))
+	c.Assert(err, gc.IsNil)
+	err = s.client.Put("/"+rurl.URL.Path()+"/meta/perm/read", []string{"bob"})
```
Comment (urosj) [inline csclient/csclient_test.go:1197]: A comment about gating maybe? Easier to detect.

**Comment by urosj** [comment]
:+1: ty!

File: csclient/csclient_test.go (hunk 2 above)
Comment (rogpeppe) [inline csclient/csclient_test.go:1197]: I just copied the code from the previous test, but you're right, they could both benefit from a comment.

File: csclient/csclient.go (hunk 1 above)
Comment (rogpeppe) [inline csclient/csclient.go]: This code has just moved from charmcmd without any changes, but I agree. I'll add a Notef.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/902/
Test PASSed.

File: csclient/csclient.go [hunk 3]
```diff
@@ -192,6 +193,22 @@ func (c *Client) UploadBundle(id *charm.Reference, b charm.Bundle) (*charm.Refer
 // entity reference. The given id should include the series and should not
 // include the revision.
 func (c *Client) uploadArchive(id *charm.Reference, body io.ReadSeeker, hash string, size int64) (*charm.Reference, error) {
+	// When uploading archives, it can be a problem that the
+	// an error response is returned while we are still writing
+	// the body data.
+	// To avoid this, we log in first so that we don't need to
+	// do the macaroon exchange after POST.
+	// Unfortunately this won't help matters if the user is logged in but
```
Comment (mhilton) [inline csclient/csclient.go:201]: It also won't help if the charm already exists in the charmstore.

File: csclient/csclient.go [hunk 4]
```diff
@@ -531,3 +548,25 @@ func (cs *Client) Log(typ params.LogType, level params.LogLevel, message string,
 	resp.Body.Close()
 	return nil
 }
+
+// Login explicitly obtains authorization credentials
+// for the charm store and stores them in the client's
+// cookie jar.
+func (cs *Client) Login() error {
```
Comment (mhilton) [inline csclient/csclient.go:555]: It would be nice if it looked for an existing cookie first, so that it doesn't login unnecessarily.

**Comment by mhilton** [comment]
:+1: with a couple of comments.

File: csclient/csclient.go (hunk 4 above)
Comment (urosj) [inline csclient/csclient.go:555]: Good catch and suggestion!

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://localhost:8080/job/charmstore-merge

**Pull request merged**